### PR TITLE
DataFrameGroupby.boxplot fails when subplots=False

### LINKED ISSUE #16748: DataFrameGroupBy.boxplot with subplots=False fails when using column param
#### Code Sample, a copy-pastable example if possible

```python
df = pd.DataFrame({'cat':np.random.choice(list('abcde'), 100), 
                    'v':np.random.rand(100), 
                    'v1':np.random.rand(100)})
df.groupby('cat').boxplot(subplots=False, column='v')
```
outputs
`KeyError: "['v'] not in index"`
#### Problem description
The boxplot works when either `subplots=False` or `column='v'` but not when they are both specified.

#### Expected Output
A single axes plot with each group having its own boxplot. The column 'cat' would label the x-axis.
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.2
pytest: 3.0.7
pip: 9.0.1
setuptools: 35.0.2
Cython: 0.25.2
numpy: 1.13.0
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: 1.5.5
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.0
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: 0.3.0.post

</details>

### PR DESCRIPTION
**Pull request by charlesdong1991**
- [x] closes #16748
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Screen Shot 2020-10-13 at 8 06 48 PM

### CODE REVIEW

**Comment by charlesdong1991** [comment]
is there a flaky test? 🤔

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -169,6 +169,7 @@ Plotting
 - Bug in :meth:`DataFrame.plot` producing incorrect legend markers when plotting multiple series on the same axis (:issue:`18222`)
 - Bug in :meth:`DataFrame.plot` when ``kind='box'`` and data contains datetime or timedelta data. These types are now automatically dropped (:issue:`22799`)
 - Bug in :meth:`DataFrame.plot.line` and :meth:`DataFrame.plot.area` produce wrong xlim in x-axis (:issue:`27686`, :issue:`25160`, :issue:`24784`)
+- Bug in :meth:`DataFrameGroupby.boxplot` when ``subplots=False``, a KeyError would raise (:issue:`16748`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.0.0.rst]: Is DataFrameGroupby user-facing?

File: pandas/plotting/_matplotlib/boxplot.py [hunk 2]
```diff
@@ -314,10 +315,21 @@ def plot_group(keys, values, ax):
             with plt.rc_context(rc):
                 ax = plt.gca()
         data = data._get_numeric_data()
+
+        # if columns is None, use all numeric columns of data; if data columns
+        # is multiIndex, which means a groupby has been applied before, select
```
Comment (gfyoung) [inline pandas/plotting/_matplotlib/boxplot.py]:
```suggestion
        # is MultiIndex, which means a GroupBy has been applied before, select
```

File: doc/source/whatsnew/v1.0.0.rst (hunk 1 above)
Comment (charlesdong1991) [reply doc/source/whatsnew/v1.0.0.rst]: my bad, changed to `DataFrame.groupby.boxplot`

File: pandas/plotting/_matplotlib/boxplot.py (hunk 2 above)
Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/boxplot.py]: changed!

**Comment by charlesdong1991** [comment]
would appreciate a lot if someone could give some hint of why the failure happens at `test_converter.py`.

**Comment by charlesdong1991** [comment]
Would appreciate a lot if you could give any hint on how `test_converter` could fail the tests? @TomAugspurger  i tried to set it to `xfail` and still get one fail (looking like a flaky test though)

File: pandas/tests/plotting/test_converter.py [hunk 3]
```diff
@@ -28,6 +28,7 @@
 pytest.importorskip("matplotlib.pyplot")
 
 
+@pytest.mark.skip(reason="There is no warnings in some versions.")
```
Comment (TomAugspurger) [inline pandas/tests/plotting/test_converter.py]: We'll need to remove this skip. Any idea why it was failing?

Do you see the warning locally when you're on this branch?

Comment (charlesdong1991) [reply pandas/tests/plotting/test_converter.py]: @TomAugspurger I tried to understand why it failed, but could find a correlation with what I changed in `test_boxplot_method.py`, In the meantime, I could not see any warning executing this function, with `test -vv` it shows there is no warning, so returns empty. And I open a notebook and run the code on this branch, i also could not see any warning. Do you have any idea why this should raise a warning? this code looks okay to me

Screen Shot 2019-08-30 at 7 53 33 PM

Comment (charlesdong1991) [reply pandas/tests/plotting/test_converter.py]: I am not sure if it is because of `backend` setup change?

Comment (TomAugspurger) [reply pandas/tests/plotting/test_converter.py]: I'm also not sure why it would have changed.

> I am not sure if it is because of backend setup change?

Do you mean a different backend in the notebook? I wouldn't expect that to matter.

Comment (charlesdong1991) [reply pandas/tests/plotting/test_converter.py]: this is where the warning defined: https://github.com/pandas-dev/pandas/blob/e55b6980fd5c99087d0959952b22e74434e9f5e2/pandas/plotting/_matplotlib/converter.py#L89
i remembered there was a PR about this merged before 0.25.1 was released, lemme check it out @TomAugspurger

Comment (charlesdong1991) [reply pandas/tests/plotting/test_converter.py]: seems this was mentioned in this PR, but very likely not relevant though. @TomAugspurger 

https://github.com/pandas-dev/pandas/pull/27481

Comment (TomAugspurger) [reply pandas/tests/plotting/test_converter.py]: Can you remove it and see if CI fails?

Comment (charlesdong1991) [reply pandas/tests/plotting/test_converter.py]: yes, removed and seems still this test failed @TomAugspurger

Comment (charlesdong1991) [reply pandas/tests/plotting/test_converter.py]: yes, removed and seems still this test failed @TomAugspurger

File: pandas/plotting/_matplotlib/boxplot.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 from pandas.core.dtypes.missing import remove_na_arraylike
 
 import pandas as pd
+from pandas import Index, IndexSlice, MultiIndex
```
Comment (TomAugspurger) [inline pandas/plotting/_matplotlib/boxplot.py]: What if you remove this import, and just use `pd.Index`, etc. below? Can you see if the test still fails?

Comment (charlesdong1991) [reply pandas/plotting/_matplotlib/boxplot.py]: wow, genius!! the issue is gone!! 👍  do you mind explaining a bit the difference? @TomAugspurger

Comment (TomAugspurger) [reply pandas/plotting/_matplotlib/boxplot.py]: No idea, unfortunately :/ Just a hunch since that test deals with first-time imports, and we were importing pandas twice. That perhaps indicates an issue with how or warning is implemented though, but that's a separate issue.

**Comment by TomAugspurger** [comment]
Just to make sure I understand, is there any API change here for an end-user? Is this code called indirectly via `DataFrame.boxplot`? I want to make sure there are no changes for that, when the user is providing a MultiIndexed dataframe and `columns`.

**Comment by charlesdong1991** [comment]
You are absolutely right, this indeed caused a change on MultiIndexed dataframe and columns when using `DataFrame.boxplot`. Sorry for not considering very comprehensively, will be more careful next time. I changed the logic a bit and test on almost all possible cases to check if the change is correct. And it looks good to me and I attach the result.

In the meanwhile, due to this, i found out actually `DataFrame.boxplot` with multiindex Dataframe are not tested at all at pandas, otherwise, this should raise an error. So I also add another test for this case, the test is exactly the same as I showed in the picture below.

Thanks again for your careful and thorough thought on the potential consequences, I am looking forward to further reviews. @TomAugspurger 

Screen Shot 2019-08-31 at 10 11 07 PM

Screen Shot 2019-08-31 at 10 11 18 PM

**Comment by TomAugspurger** [comment]
> But with the new change, this can be plotted, and I feel this change is more useful and users might like ti.

The alternative would be for the user to provide something like `columns=pd.IndexSlice[:, 'two']`, correct?

Just to make sure, there's no change to the case where the user-provided `columns` is from the outer level like `bar`?

My main concern is that by automatically doing this, we'd introduce some kind of ambiguity about what `columns` means.

**Comment by charlesdong1991** [comment]
sorry for my late response, was out sick for a couple days. I read your comments again and think I fully get your point. Indeed, this will also change the current behavior from the outer level like `bar`. Right now, if doing `df.boxplot(columns='bar')`, will get:

Screen Shot 2019-09-05 at 8 35 44 PM

You mean we should keep this behavior, right? @TomAugspurger 

And I also have a second thought, and think we might provide another argument for boxplot, which is something like `include` for multiindex columns, if `True`, then users could define more ambiguous columns to get their columns plotted, so `columns='bar'` will plot all columns as long as `'bar'` is in any multi-level index.

**Comment by jreback** [comment]
@charlesdong1991 can you merge master; can you summarize the issues here that were raised?

**Comment by jreback** [comment]
can you merge master

**Comment by charlesdong1991** [comment]
The original issue was, in the case such as:
```python
df = pd.DataFrame({'cat':np.random.choice(list('abcde'), 100), 
                    'v':np.random.rand(100), 
                    'v1':np.random.rand(100)})
df.groupby('cat').boxplot(subplots=False, column='v')
```
Since after groupby, this `v` will not be in the column index, so will raise an error. But with `pd.IndexSlice[:, column]`, seems issue could be solved for `groupby.boxplot`. However, @TomAugspurger brought up a very important point that this will however have an endpoint change for normal `df.boxplot` if specified column comes from the first level in column index (for MultiIndex case).

I just updated my previous solution, and now it won't afffect endpoint change for `df.boxplot` on the first level of MultiIndex. However, when it comes to the second level, behaviour will change. Since my description might cause confusion, I will present using several graphs, so pls bear with me, the df is:
```python
arrays = [['bar', 'bar', 'baz', 'baz', 'foo', 'foo', 'qux', 'qux'],
            ['one', 'two', 'one', 'two', 'one', 'two', 'one', 'two']]
tuples = list(zip(*arrays))
index = pd.MultiIndex.from_tuples(tuples, names=['first', 'second'])
df = pd.DataFrame(np.random.randn(3, 8), index=['A', 'B', 'C'], columns=index)
```
Current behaviour in Pandas, it will raise an error on second level:
Screen Shot 2019-10-19 at 8 21 03 PM

With this PR:
Screen Shot 2019-10-19 at 8 21 43 PM

As you could see above, the second level end behaviour will change, the rest will remain the same. And the original issue for `groupby.boxplot` can still be fixed. @jreback

**Comment by jreback** [comment]
can you merge master and we'll look again

**Comment by charlesdong1991** [comment]
Merged! @jreback 

Looks good, except for a doc error raised by `Cython` due to new release of `ipython`.

File: pandas/tests/plotting/test_boxplot_method.py [hunk 5]
```diff
@@ -440,3 +440,71 @@ def test_fontsize(self):
...
+        df = DataFrame(
+            {
+                "cat": np.random.choice(list("abcde"), 100),
+                "v": np.random.rand(100),
+                "v1": np.random.rand(100),
+            }
+        )
+        grouped = df.groupby("cat")
+
+        # check is boxplot with subplots=False works
```
Comment (WillAyd) [inline pandas/tests/plotting/test_boxplot_method.py]: Shouldn’t need these comments in the tests; they just state what is already on the next line(s)

File: pandas/plotting/_matplotlib/boxplot.py [hunk 6]
```diff
@@ -335,10 +335,25 @@ def plot_group(keys, values, ax):
             with plt.rc_context(rc):
                 ax = plt.gca()
         data = data._get_numeric_data()
-        if columns is None:
-            columns = data.columns
+
+        if columns:
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/boxplot.py]: Was the condition change here from `is None` to just testing from truthiness required? It would change behavior of `columns=False` which doesn’t look tested; probably best to leave alone

File: pandas/plotting/_matplotlib/boxplot.py [hunk 7]
```diff
@@ -335,10 +335,25 @@ def plot_group(keys, values, ax):
...
+
+            # this is to align the current behavior of boxplot on columns, so if users
+            # explicitly specify column names and it is in df columns, then take the
+            # subset; Or if it is columns is in the first level, this is set mainly to
+            # avoid API change
+            if set(columns).issubset(data.columns) or set(columns).issubset(
```
Comment (WillAyd) [inline pandas/plotting/_matplotlib/boxplot.py]: Hmm what’s the use case for just looking at the first level of the MultiIndex for label matching?

**Review by WillAyd — COMMENTED** [review]
@charlesdong1991 is this still active? If so let’s try and push through!

**Comment by charlesdong1991** [comment]
thanks a lot for picking it up @WillAyd  Yeah, I am active, so is this PR!  xD
I think there is a design issue with this PR, and now the priority is regression and new issues posted for 1.0

I will get back to your comments later next week, since I also came across an issue which requires a feature for plotting based on the Index level, might be relevant to this PR, I will try to get that feature done and then back to address and try to push it through

**Comment by WillAyd** [comment]
Can you move the whatsnew?

**Comment by simonjayhawkins** [comment]
@charlesdong1991 what's the status here?

**Comment by charlesdong1991** [comment]
yeah, i will give this PR another try later this week @simonjayhawkins

**Comment by jreback** [comment]
whats the status here?

**Comment by jbrockmendel** [comment]
@charlesdong1991 gentle ping can you rebase

**Comment by charlesdong1991** [comment]
hey @jreback @WillAyd @simonjayhawkins @TomAugspurger 

sorry for very late update here. I just took a look at the original issue and my past commits together with the relevant codebase, I feel it was quite inefficient and confusing. So here, I flipped most of the changes and I _think_ now it looks much better.

So the issue was with a df like below, it will fail doing `df.groupby.boxplot(subplots=False, column='v')`:
```python
df = pd.DataFrame({'cat':np.random.choice(list('abcde'), 100), 
                    'v':np.random.rand(100), 
                    'v1':np.random.rand(100)})
df.groupby('cat').boxplot(subplots=False, column='v')
```
and it is because the data for plotting after `df.groupby('cat')` has changed to MI, so `v` no longer exists in the transformed data.

Therefore, the new solution for this is quite simple, I couple the keys of groupby values (in this case, [a, b, c, d, e]) with the column value sellers assign to `column` argument (in this case, `v`), so we have `[(a, v), (b,v), (c, v), (d, v), (e, v)]` and assign them to `boxplot` function, and then `boxplot` function will look for subset based on this new `column` values, instead of `v` in the original df which is used by sellers.

How does it sound now? Feedbacks and reviews are very welcomed!

Regarding the error, It seems irrelevant to this PR.

**Comment by jreback** [comment]
@charlesdong1991 this looks good, can you post an updated plot (and your tests look nice); merge master and ping on green-ish..

**Comment by charlesdong1991** [comment]
Screen Shot 2020-10-13 at 8 06 48 PM

sure, these are plots with and without specifying `column`.

**Comment by jreback** [comment]
@charlesdong1991 can you merge master

**Comment by charlesdong1991** [comment]
small ping here @jreback sorry for the late and long response

**Comment by jreback** [comment]
thanks @charlesdong1991 very nice!
